"How we handled Log4J using SCB" Blog Post (#226)
secureCodeBox/documentation: blog/2022-01-18-log4shell.md

---
# SPDX-FileCopyrightText: 2021 iteratec GmbH
#
# SPDX-License-Identifier: Apache-2.0

title: How We Used the secureCodeBox In Our Log4Shell Incident Response
author: The secureCodeBox Team
author_title: Core Developer
author_url: https://github.com/secureCodeBox
author_image_url: https://avatars.githubusercontent.com/u/34573705?v=4
tags:
  - secureCodeBox
  - case study
  - incident response
  - SAST
  - DAST
  - cascading scans
  - Log4Shell
  - Log4J
description: We describe how the secureCodeBox allowed us to rapidly respond to the Log4shell vulnerability.
image: /img/blog/2022-02-24-burninglog.jpg
---

![A burning log](/img/blog/2022-02-24-burninglog.jpg)

Cover photo by [Ray Shrewsberry](https://unsplash.com/@ray12119) on [Unsplash](https://unsplash.com/photos/AXRYB88LAXs).

By now, you must have heard about Log4Shell, the present that ruined Christmas for many developers and IT specialists, whether naughty or nice. 

> This blog describes how we used the secureCodeBox in our incident response process at Iteratec.

Reviewer: ```suggestion
This blog describes how we used the secureCodeBox in our incident response process at iteratec.
```

   
<!--truncate-->

## A Brief Introduction To Log4Shell

> But first, a small refresher: In late November 2021, a [zero-day vulnerability was discovered in the widely used Java logging library Log4J](https://nakedsecurity.sophos.com/2021/12/13/log4shell-explained-how-it-works-why-you-need-to-know-and-how-to-fix-it/). It allowed attackers to remotely execute code through Java Naming Directory Index ([JNDI](https://en.wikipedia.org/wiki/Java_Naming_and_Directory_Interface)) lookups to malicious LDAP servers: If an attacker can get the application to log a payload controlled by the attacker, like `{jndi:ldap//evil.ldap.server.adress/a}`, then the code hosted on the LDAP server would be loaded and executed by the program, effectively letting third parties take control of the Java Application and the server it's running on.

Reviewer 2: ```suggestion
But first, a small refresher: In late November 2021, a [zero-day vulnerability was discovered in the widely used Java logging library Log4J](https://nakedsecurity.sophos.com/2021/12/13/log4shell-explained-how-it-works-why-you-need-to-know-and-how-to-fix-it/). It allowed attackers to remotely execute code through Java Naming and Directory Interface ([JNDI](https://en.wikipedia.org/wiki/Java_Naming_and_Directory_Interface)) lookups to malicious LDAP servers: If an attacker can get the application to log a payload controlled by the attacker, like `{jndi:ldap//evil.ldap.server.adress/a}`, then the code hosted on the LDAP server would be loaded and executed by the program, effectively letting third parties take control of the Java Application and the server it's running on.
```

This vulnerability shook the IT world. It received a CVE rating of 10/10, and even the German government issued a statement calling for immediate action and described the issue as "critical". This is due to two main reasons. First, the vulnerability was relatively simple for attackers to exploit. Second, it has remained undiscovered since 2013, affecting many services from AWS to Minecraft.

> At [Iteratec](https://www.iteratec.com), as a software development company, we had to assess our security posture as well - for both the infrastructure that we were running for ourselves, as well as the software we develop for our customers.

Reviewer: ```suggestion
At [iteratec](https://www.iteratec.com), as a software development company, we had to assess our security posture as well - for both the infrastructure that we were running for ourselves, as well as the software we develop for our customers.
```

We use demo targets in SCB to continuously test the functionality of our scanners during the development cycles.
Our existing demo targets include the [bodgeit store](https://github.com/psiinon/bodgeit) and [OWASP Juice Shop](https://owasp.org/www-project-juice-shop/).  
Integration tests are run against the demo targets during our CI/CD pipeline to spot any malfunctioning scanner.

Demo targets consist of a Kubernetes service containing a vulnerable application image.
Creating a new Log4J demo target is a straightforward process.

First, a vulnerable docker image is required. 
Luckily, an image has been [provided by the GitHub user 'christophetd'](https://github.com/christophetd/log4Shell-vulnerable-app).
Second, we must create our [helm chart folder](https://github.com/secureCodeBox/secureCodeBox/tree/main/demo-targets/vulnerable-log4j/templates) using the same directory structure as the other demo targets and configure it to use the vulnerable Log4J image.
We set the `image.repository` to the provided docker image above as seen in the following `values.yaml` file:

```yaml

> replicaCount: 1

Reviewer: I'm not quite sure if or why it is helpful to explain the process how to "implement" a specific demo target 🤔 What is in it for me as reader (how is not yet a contributer)?

Author: The idea is that we show the entire process for someone if they want to do incident response. That way, they also know how to build their own demo target to test their detection rules against.

      containers:
        - name: {{ .Chart.Name }}
          securityContext:
            {{- toYaml .Values.securityContext | nindent 12 }}
          image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
          imagePullPolicy: {{ .Values.image.pullPolicy }}
          ports:
            - name: http
              containerPort: 8080
              protocol: TCP
          resources:
            {{- toYaml .Values.resources | nindent 12 }}
      {{- with .Values.nodeSelector }}
      nodeSelector:
        {{- toYaml . | nindent 8 }}
      {{- end }}
    {{- with .Values.affinity }}
      affinity:
        {{- toYaml . | nindent 8 }}
    {{- end }}
    {{- with .Values.tolerations }}
      tolerations:
        {{- toYaml . | nindent 8 }}
    {{- end }}
```

And then we are essentially done. All that is left now is to install the demo target in the preferred namespace, as shown below:  

```bash
helm upgrade --install vulnerable-log4j ./demo-targets/vulnerable-log4j/ --namespace <NAMESPACE>
```

After using the demo target to validate that the scanner works, we then proceeded to run it against our own infrastructure.

### Finding Hosts

> To scan for the Log4Shell vulnerabilities dynamically we first have to identify what to scan.

Reviewer: ```suggestion
To scan for the Log4Shell vulnerabilities dynamically (DAST) we first have to identify what to scan, the so called attack surface.
```

> Using the secureCodeBox we used the [cascading scanning mechanism](https://docs.securecodebox.io/docs/hooks/cascading-scans) to run an initial scan to discover scan targets and use their result to automatically start scans for the identified hosts and domains.

Reviewer: ```suggestion
One powerful feature of the secureCodeBox is the dynamic scan orchestration of different security scanner:
 the [cascading scans mechanism](https://docs.securecodebox.io/docs/hooks/cascading-scans).  Based on that it is easy to run an initial scan to discover scan targets and use their result to automatically start (cascade) specialized scans for the identified hosts and domains.
```

We used two different different discovery methods:

> 1. For hosts in the internal networks, we used the secureCodeBox `nmap` scanType to scan for hosts and open ports in our internal IP ranges (e.g. 10.42.0.0/16). Every port which nmap identified to be related to http(s) (which is generally the easiest protocol to scan for Log4Shell even thought it can also be exploitable via different protocols) was used as a target in a Log4Shell scans.

Reviewer: ```suggestion
1. For hosts sitting in internal networks, we used nmap ([with the secureCodeBox `nmap` scanType](https://www.securecodebox.io/docs/scanners/nmap)) to identify active hosts and open ports in our internal IP ranges (e.g. 10.42.0.0/16) and network segments. Every port which nmap identified to be related to http(s) (which is generally the easiest protocol to scan for Log4Shell even thought it can also be exploitable via different protocols) was used as a target in a cascading Log4Shell scan.
```

> 2. For publicly available hosts, we used the `amass` scanType to find subdomains for the list of domain names we own as a company. This outputs a list of a subdomains which we then use to automatically trigger `nmap` scans to find open http(s) ports for the actual Log4Shell vulnerability assessment.

Reviewer: ```suggestion
2. For publicly available hosts, we used the OWASP AMASS scanner ([with the secureCodeBox `amass` scanType](https://www.securecodebox.io/docs/scanners/amass)) first to find subdomains for the list of domain names we own as a company. This outputs a list of a subdomains which also automatically trigger `nmap` cascading scans to find open http(s) ports for the actual Log4Shell vulnerability assessment.
```

          service: "https"
          state: open
      - category: "Open Port"
        attributes:
          service: "https*"
          state: open
  scanSpec:
    scanType: "nuclei"
    # This example uses the ConfigMap we created further up in the article.
    # If you want to use the official set of Nuclei templates for Log4J,
    # change the parameterization below as previously described.
    parameters:
      - "-templates"
      - "/custom-nuclei-rules/log4j-template.yaml"
      - "-target"
      # Target domain name of the finding and start a nuclei scan
      - "https://{{$.hostOrIP}}:{{attributes.port}}"
    volumeMounts:
      - name: nuclei-template-log4j
        mountPath: /custom-nuclei-rules/
    volumes:
    - name: nuclei-template-log4j
      configMap:
        name: nuclei-template-log4j
```

## Finding Affected Code

Of course, as a software development company, we also had to validate that the code we produce for our customers wasn't affected.
The individual development teams quickly determined if their projects were affected, created updates, and shipped them to the customers.
As part of the security team, we supported the teams in their efforts.

> In parallel, we used the static code analysis capabilities of the secureCodeBox to scan our software repositories for places where code may have been missed.

Reviewer: ```suggestion
In parallel, we used the static code analysis capabilities (SAST) of the secureCodeBox to scan our software repositories for places where code may have been missed.
```
